### PR TITLE
📖  Multitenancy: Restrict controller credential usage by creating a permission resource

### PR DESCRIPTION
**Pull request by sedefsavas**
**What this PR does / why we need it**:
Currently, we don't provide a mechanism to restrict the usage of controller credentials. If an AWSCluster is created without `principalRef`, by default, CAPA's credentials will be used to create the resources in that cluster. This contradicts with `FR4` which is preventing privilege escalations.

Since we cannot make `principalRef` compulsory due to old clusters (that was created before multitenancy), we should find a way to restrict access to CAPA credentials from AWSClusters that should not access it.

One option would be to introduce a new permission CRD that will be created in `capa-system`, which will be required to use CAPA AWS credentials. Users that can create that permission resource in capa-system, will be eligible to use CAPA credentials.

I also described a workaround during upgrade by creating permission resources automatically for existing clusters.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by sedefsavas** [comment]
Same issue is also present in CAPZ. cc @CecileRobertMichon @nader-ziada @devigned

**Comment by devigned** [comment]
I would rather deprecate the use of the controller identity for provisioning clusters. I think it makes more sense to have the resource be specific about the identity it is using to reconcile the cluster.

**Comment by sedefsavas** [comment]
@devigned This will make playing with the providers more difficult for new users IMO.

**Comment by devigned** [comment]
CAPZ requires no Azure access to be deployed. The controller doesn't need an identity until it deploys a cluster. 

Requiring a user to provide an identity to deploy the controller is more difficult than asking them to provide an identity to reconcile a cluster (IMO). Also, having more complex fallback logic is also more difficult. 

Additional to that, what is the behavior when you move a cluster to a different management cluster? The cluster is not fully described by it's resource. It now has a dependency on the controller environment vars which might be different in the target management cluster.

All in all, I think the env vars are vestigial, confusing and should be deprecated, but I'm open to be convinced otherwise.

**Comment by sedefsavas** [comment]
> I would rather deprecate the use of the controller identity for provisioning clusters.
For existing clusters, during upgrade, we should figure out how to handle not using controller identity. I don't see an easy way to enforce using identity's for the existing clusters.

**Comment by sedefsavas** [comment]
> All in all, I think the env vars are vestigial, confusing and should be deprecated, but I'm open to be convinced otherwise.

Instead of creating a permanent object, we can create and delete a say `Secret` in the controller namespace to assume that the user is eligible to use controller credentials.
This sounded hacky to me, so instead suggested a new CRD so that we can also see which clusters are using controller credentials. But since there are concerns regarding dealing with a new var, this may also work.

**Comment by CecileRobertMichon** [comment]
In CAPZ, we're thinking of keeping the controller identity as an optional (it's required in v1alpha3) deprecated fallback method in v1alpha4 but document using per cluster identity as best practice (and controller vars as deprecated). Then, in the next API version (alpha5 or beta1), we will completely remove controller identity vars.

File: docs/proposal/20200506-single-controller-multitenancy.md [hunk 1]
```diff
@@ -538,6 +537,46 @@ spec:
     name: test-account-creds
 ```
 
+<strong><em>Permission CRD</em></strong>
+
+<em>ControllersPrincipalUsagePermission</em>
+
+When `principalRef` is not provided in AWSCluster, controller falls back to use controller credentials, 
+but those credentials tends to have more comprehensive permissions than workload user would need.
```
Comment (randomvariable) [inline docs/proposal/20200506-single-controller-multitenancy.md:545]: `but those credentials tends to have more comprehensive permissions than workload user would need.`

I don't think that's strictly true - the user still doesn't have direct access to the credentials, I think the more accurate statement is along the lines of:

the controller's credentials are necessarily tied to an AWS account, and in many cases this is where the management cluster is deployed, and the cloud administrators will not want users to provision workload clusters in the management AWS account.

**Review by randomvariable — CHANGES_REQUESTED** [review]
I'm unsure about this current proposal as it uses a different RBAC model. I've suggested an alternative that fits the current behaviour and also makes explicit for each AWSCluster what credentials are being used.

**Comment by randomvariable** [comment]
> All in all, I think the env vars are vestigial, confusing and should be deprecated, but I'm open to be convinced otherwise.

For AWS, it's not even env vars. It's the AWS SDK default credential provider chain which isn't going away (multi-account access may still use the default chain for role assumption across accounts and the source maybe EC2 instance profiles, kiam or K8s service principals coming via IRSA pod webhook injection), for example. Maybe less important for CAPZ.

**Comment by devigned** [comment]
> Maybe less important for CAPZ.

We have the same thing going on with Azure SDK's auth chain. At this point, we specify which fallback to use based on the identity ID, or lack of one, for a given cluster. When default credentials are deprecated, we'll specify the specific identity for each cluster and omit any fallback to other credential providers.

I thought you can do the same with AWS, but you'd know better than I.

**Comment by randomvariable** [comment]
It's the use of "ambient" credentials to springboard access to another AWS account via AWS*RolePrincipal that we need to preserve ( https://docs.aws.amazon.com/IAM/latest/UserGuide/tutorial_cross-account-with-roles.html ). I don't think Azure's IAM model works like that from previous discussions.

File: docs/proposal/20200506-single-controller-multitenancy.md [hunk 2]
```diff
@@ -538,6 +537,46 @@ spec:
...
+To prevent privilege escalation as per [FR4](#FR4), we intend to restrict the usage of controller credentials by AWSClusters which
+has a `ControllersPrincipalUsagePermission` resource in the controllers namespace. If workload operator does not have enough privileges to create this object in the controller namespace or
```
[the remark this answers is not in the document]
Comment (sedefsavas) [reply docs/proposal/20200506-single-controller-multitenancy.md:547]: I might be missing something but how will this prevent an AWSCluster that is created by a workload user to use controller principal?

[the remark this answers is not in the document]
Comment (sedefsavas) [reply docs/proposal/20200506-single-controller-multitenancy.md]: Discussed offline.
This model supports restricting access to the controller credentials in the namespace level, which is consistent with the rest of the proposal.
Updated the proposal accordingly.

File: docs/proposal/20200506-single-controller-multitenancy.md [hunk 3]
```diff
@@ -845,8 +884,15 @@ different namespaces do not break the AWS Cloud Provider.
 ## Upgrade Strategy
 
 The data changes are additive and optional, so existing AWSCluster specifications will continue to reconcile as before.
-These changes will only come into play when an IAM role is provided in the new field in AWSClusterSpec. Upgrades to
-versions with this new field will be broken
+These changes will only come into play when an IAM role is provided in the new field in AWSClusterSpec.
+During v1alpha3 to v1alpha4 upgrade, not to break existing clusters, 
+`ControllersPrincipalUsagePermission` object will be automatically created (by controller) in the controller namespace for `ready` AWSClusters.
+For all other clusters, if `principalRef` is nil, then `ControllersPrincipalUsagePermission` will be required to be created by either workload or management user.
+To differentiate AWSClusters that initially had `principalRef` (and `ready`) with the old clusters
```
[the remark this answers is not in the document]
Comment (sedefsavas) [reply docs/proposal/20200506-single-controller-multitenancy.md]: This is no longer an issue with the changes.

**Comment by sedefsavas** [comment]
Updated the proposal following @randomvariable's comments. 
Now there is a singleton resource (AWSClusterControllerPrincipal) that guards the usage of controller credentials from all namespaces.

**Comment by randomvariable** [comment]
Thanks!

/lgtm

/hold
for consensus from next monday's meeting.

@sedefsavas It'll be good to get more eyes on this, but please don't let this stop you from starting implementation.

**Pull request merged**